### PR TITLE
Add support for custom JinjaTracer implementations

### PR DESCRIPTION
**Pull request by james-johnston-thumbtack**
### Brief summary of the change made

Currently, the JinjaAnalyzer implementation will directly construct an instance of JinjaTracer, making it difficult for users providing a derived JinjaAnalyzer to also have it return a derived JinjaTracer.

By moving the construction of JinjaTracer to a separate function, it is easy for plugin authors to override this function and provide their own tracer.

This approach is very similar to the one taken for allowing custom JinjaAnalyzer implementations in pull request #5543.

### Are there any other side effects of this change that we should be aware of?

No.  This is a small refactor that makes it easier to write derived classes.

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/templaters/slicers/tracer.py [hunk 1]
```diff
@@ -51,6 +52,7 @@ def __init__(
         raw_slice_info: Dict[RawFileSlice, RawSliceInfo],
         sliced_file: List[TemplatedFileSlice],
         render_func: Callable[[str], str],
+        config: Optional[FluffConfig] = None,
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/templaters/slicers/tracer.py:55]: Do we need to pass the whole config in here? I'm not necessarily against it, but it seems unnecessary at the moment. Are you doing something interesting with it in the derived case?

Comment (james-johnston-thumbtack) [reply src/sqlfluff/core/templaters/slicers/tracer.py:55]: Actually no.... not currently using it in the derived case (and don't currently expect to need to).  This was more of a future-proofing move.  I can remove it.

Comment (james-johnston-thumbtack) [reply src/sqlfluff/core/templaters/slicers/tracer.py]: Added on a commit that removes the config parameters, including from the _get_jinja_analyzer that I added earlier this year but don't really need it there either.

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**